### PR TITLE
Added documentation for tests:check-direct-dependency-use command, #PG-3272

### PR DESCRIPTION
**Pull request by AltamashShaikh**
### Description:

Added documentation for tests:check-direct-dependency-use command
Fixes: #PG-3272

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by michalkleiner** [comment]
Nice start, thanks @AltamashShaikh!

Can we add a note why this might be useful to do? Why would plugin developers want to do this and if they find using core dependencies, what they can do about it?

**Comment by AltamashShaikh** [comment]
@michalkleiner Can you check now ?

File: docs/5.x/tests-system.md [hunk 1]
```diff
@@ -195,6 +195,44 @@ class ApiTest extends SystemTestCase
 }
 ```
 
+## How to write a test to scan for Matomo core dependencies being used directly in your plugins
+
+With the release of Matomo `v5.1.0`, you can now check the usage of core dependencies in your plugin directly using the `tests:check-direct-dependency-use` command. With the release of Matomo 5, plugins should not use core dependencies directly but instead [prefix](https://developer.matomo.org/guides/migrate-matomo-4-to-5#vendor-proxies) them. You can also write a system test case to test the same by using the sample code below
```
Comment (michalkleiner) [inline docs/5.x/tests-system.md]:
```suggestion
With the release of Matomo `5.1.0`, you can now easily check if your plugin is using any core dependencies by running the `tests:check-direct-dependency-use` command. With the release of Matomo 5, plugins should not use core dependencies directly but instead use them via [proxies](https://developer.matomo.org/guides/migrate-matomo-4-to-5#vendor-proxies). To have a check for this in your test suite and run it automatically, you can also write a system test case by using the sample code below:
```

File: docs/5.x/tests-system.md [hunk 2]
```diff
@@ -195,6 +195,44 @@ class ApiTest extends SystemTestCase
...
+
+```php
+<?php
+
+use Piwik\Plugins\TestRunner\Commands\CheckDirectDependencyUse;
```
Comment (michalkleiner) [inline docs/5.x/tests-system.md]:
```suggestion
```php
use Piwik\Plugins\TestRunner\Commands\CheckDirectDependencyUse;
```

Comment (michalkleiner) [reply docs/5.x/tests-system.md]: I checked other examples and we don't use the opening tag in them.

File: docs/5.x/tests-system.md [hunk 3]
```diff
@@ -195,6 +195,44 @@ class ApiTest extends SystemTestCase
...
+use Piwik\Tests\Framework\TestCase\SystemTestCase;
+use Piwik\Version;
+use Symfony\Component\Console\Input\ArrayInput;
+use Symfony\Component\Console\Output\NullOutput;
+
+class CheckDirectDependencyUseCommandTest extends SystemTestCase
+{
+    public function testCommand()
+    {
+        if (version_compare(Version::VERSION, '5.0.2', '<=') && !\Piwik\file_exists(PIWIK_INCLUDE_PATH . '/plugins/TestRunner/Commands/CheckDirectDependencyUse.php')) {
```
Comment (michalkleiner) [inline docs/5.x/tests-system.md]: Should this test for versions before 5.1.0 or when the file is missing like so?

```suggestion
        if (version_compare(Version::VERSION, '5.1.0', '<') || !\Piwik\file_exists(PIWIK_INCLUDE_PATH . '/plugins/TestRunner/Commands/CheckDirectDependencyUse.php')) {
```

File: docs/5.x/tests-system.md [hunk 4]
```diff
@@ -195,6 +195,44 @@ class ApiTest extends SystemTestCase
...
+        $checkDirectDependencyUse = new CheckDirectDependencyUse();
+        $console->addCommands([$checkDirectDependencyUse]);
+        $command = $console->find('tests:check-direct-dependency-use');
+        $arguments = array(
+            'command'    => 'tests:check-direct-dependency-use',
+            '--plugin' => $pluginName,
+            '--grep-vendor'
+        );
+        $inputObject = new ArrayInput($arguments);
+        $command->run($inputObject, new NullOutput());
```
Comment (michalkleiner) [inline docs/5.x/tests-system.md]: Just a bit of formatting tweak for easier reading of the code.

```suggestion
        }
        
        $pluginName = '{YOUR_PLUGIN_NAME}';
        
        $checkDirectDependencyUse = new CheckDirectDependencyUse();
        $console = new \Piwik\Console(self::$fixture->piwikEnvironment);
        $console->addCommands([$checkDirectDependencyUse]);
        
        $command = $console->find('tests:check-direct-dependency-use');
        $arguments = [
            'command'  => 'tests:check-direct-dependency-use',
            '--plugin' => $pluginName,
            '--grep-vendor',
        ];
        $inputObject = new ArrayInput($arguments);
        $command->run($inputObject, new NullOutput());
```

**Comment by michalkleiner** [comment]
Generally looking ok, just some minor tweaks/suggestions.
We will merge this when 5.1.0-beta1 is released.

**Comment by AltamashShaikh** [comment]
@michalkleiner Applied all the feedback.

**Comment by michalkleiner** [comment]
Thanks, I guess we shall wait for 5.1.0-beta1 release and then we can merge this documentation PR.
